src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15799,34 +15801,208 @@ def stack(
+            )
+        if not sort:
+            ErrorMessage.not_implemented(
+                "Snowpark pandas DataFrame/Series.unstack does not yet support the `sort` parameter"
             )
-            .sort_rows_by_column_values(
-                columns=index_names,  # type: ignore
+        if self._modin_frame.is_multiindex(axis=1):
+            ErrorMessage.not_implemented(
+                "Snowpark pandas doesn't support multiindex columns in the unstack API"
+            )
+
+        level = level if is_list_like(level) else [level]
+
+        index_names = self.get_index_names()
+
+        # Check to see if we have a MultiIndex, if we do, make sure we remove
+        # the appropriate level(s), and we pivot accordingly.
+        if len(index_names) > 1:
+            # Resetting the index keeps the index columns as the first n data columns
+            qc = self.reset_index()
+            index_cols = qc._modin_frame.data_column_pandas_labels[0 : len(index_names)]
+            pivot_cols = [index_cols[lev] for lev in level]  # type: ignore
+            index_cols = [
+                index_cols[i]
+                for i in range(len(index_cols))
+                if index_cols[i] not in pivot_cols
+            ]
+            vals = [
+                c for c in self.columns if c not in index_cols and c not in pivot_cols
+            ]
+            # We need to track where the index and columns originally had no name in order to reset
+            # the those names back to None after the operation
+            unnamed_index_positions = [
+                i for i, x in enumerate(index_names) if x is None
+            ]
+            column_names_to_reset_to_none = []
+            # Track the new column names for the original unnamed index
+            for i in unnamed_index_positions:
+                column_names_to_reset_to_none.append(
+                    qc._modin_frame.data_column_pandas_labels[i]
+                )
+
+            qc = qc.pivot_table(
+                columns=pivot_cols,
+                index=index_cols,
+                values=vals,
+                aggfunc="min",
+                fill_value=fill_value,
+                margins=False,
+                dropna=True,
+                margins_name="All",
+                observed=False,
+                sort=sort,
+            )
+
+            # Set the original unnamed index values back to None
+            output_index_names = qc.get_index_names()
+            output_index_names_replace_level_with_none = [
+                None
+                if output_index_names[i] in column_names_to_reset_to_none
+                else output_index_names[i]
+                for i in range(len(output_index_names))
+            ]
+            qc = qc.set_index_names(output_index_names_replace_level_with_none)
+            # Set the unnamed column values back to None
+            output_column_names = qc.columns.names
+            output_column_names_replace_level_with_none = [
```

Comment:
is the reset of column needed, it seems that the rest of the original index column is used as the pivot columns, and the result columns name are values of from the pivot column, the original column names shouldn't be in any of the result column of pivot

Reply:
We still need this code as some of the names we are replacing with None could end up in the column names, see the case below:
```
    index = pd.MultiIndex.from_product(
        iterables=[[2, 1], ["a", "b"]], names=["hello", None]
    )
    native_df = native_pd.DataFrame(np.random.randn(4), index=index, columns=["A"])
    snow_df = pd.DataFrame(native_df)
    snow_df.unstack()
```

```
snow =                 A          
level_1         a         b
hello                      
1       -0.407852  0.038916
2        2.405507  0.374415
expected_pandas =               A          
              a         b
hello                    
1     -0.407852  0.038916
2      2.405507  0.374415
```